dali/python/nvidia/dali/plugin/base_iterator.py

```diff
@@ -0,0 +1,185 @@
+                "All pipelines readers should have the same shard number set"
+
+            self._shards_id = np.array([p.shard_id(self._reader_name) for p in self._pipes], dtype=np.int)
+
+            assert np.all([p.is_pad_last_batch(self._reader_name) for p in self._pipes]) or \
+                   not np.any([p.is_pad_last_batch(self._reader_name) for p in self._pipes]), \
+                "All pipelines readers should have set padding in the same way"
+            self._last_batch_padded = self._pipes[0].is_pad_last_batch(self._reader_name)
+
+            assert np.all([p.is_stick_to_shard(self._reader_name) for p in self._pipes]) or \
+                   not np.any([p.is_stick_to_shard(self._reader_name) for p in self._pipes]), \
+                "All pipelines readers should have set stick to the shard in the same way"
+            self._is_stick_to_shard = self._pipes[0].is_stick_to_shard(self._reader_name)
+
+            if self._last_batch_padded:
+                # if padding is enabled all shards are equal
+                self._size = self._pipes[0].epoch_size(self._reader_name, True) // self._shards_num
+            else:
+                # get the size as a multiply of the batch size that is bigger or equal than the biggest shard
+                self._size = math.ceil(math.ceil(self._size_no_pad / self._shards_num) / self.batch_size) * self.batch_size
+
+            # cont where we starts inside each GPU shard in given epoch,
+            # if shards are uneven this will differ epoch2epoch
+            self._counter_per_gpu = np.zeros(self._shards_num, dtype=np.long)
+            shards_beg = np.floor(np.arange(0, self._shards_num) * self._size_no_pad / self._shards_num).astype(np.int)
+            shards_end = np.floor(np.arange(1, self._shards_num + 1) * self._size_no_pad / self._shards_num).astype(np.int)
+            self._shard_sizes = shards_end - shards_beg
+
+    def _check_stop(self):
+        """"
+        Checks iterator stop condition and raise StopIteration if needed
+        """
+        if self._counter >= self._size and self._size > 0:
+            if self._auto_reset:
+                self.reset()
+            raise StopIteration
+
+    def _remove_padded(self):
+        """
+        Checks if remove any padded sample and how much
+        """
+        if_drop = False
+        left = -1
+        if not self._fill_last_batch:
+            # calculate each shard size for each id, and check how many samples are left by substracting
+            # from iterator counter the shard size, then go though all GPUs and check how much data needs to be dropped
+            shards_beg = np.floor(self._shards_id * self._size_no_pad / self._shards_num).astype(np.int)
+            shards_end = np.floor((self._shards_id + 1) * self._size_no_pad / self._shards_num).astype(np.int)
+            left = self.batch_size - (self._counter - (shards_end - shards_beg))
```

Comment:
Can we use the _shard_sizes from line 90?

Reply:
No, because here `_shards_id` is the id of the shard the reader is reading now, not the initial one.
Also in `90` we calculate shard sizes for every pipeline, here we care only about the one that iterator wraps up.

Reply:
I was just thinking about looking up _shard_sizes[self._shard_id] instead of computing them again.

Reply:
Done. But I don't think it helps much (usually we calculates it for 8 values at most, not we just looking up for it).